### PR TITLE
[4.0] Add fallback for not complete manifest caches used in the pre-update checker

### PR DESCRIPTION
**Pull request by zero-24**
Pull Request for Issue https://github.com/joomla/joomla-cms/pull/22421#issuecomment-531125432 reported by @infograf768 

### Summary of Changes

Add fallback for not complete manifest caches used in the pre-update checker

### Testing Instructions

.- install 4.0-dev
- install patchtester
- go to the database and clear the field `manifest_cache` for the patchtester record in the extensions table.
- set the [nightly major update server](https://update.joomla.org/core/nightlies/next_major_list.xml)
- notice the pre update checker
- notice that the pre update checker shows nothing as version for the patchtester and raises a notice
- apply this patch
- the updater now shows the string `Unknown version in this case and should not raise and notice any more.

### Expected result

No errors; After the patch the pre-update checker shows `Unknown version` before that info was just missing / empty.

### Actual result

```
Notice: Trying to get property 'version' of non-object in .../administrator/components/com_joomlaupdate/Model/UpdateModel.php on line 1357
```

And an empty version info for that extension.

### Additional steps

- [ ] [3.10-dev needs to be updated too.](https://github.com/joomla/joomla-cms/blob/3.10-dev/administrator/components/com_joomlaupdate/models/default.php#L1387)

### CODE REVIEW

File: administrator/language/en-GB/en-GB.com_joomlaupdate.ini [hunk 1]
```diff
@@ -20,6 +20,7 @@ COM_JOOMLAUPDATE_LOGIN="Login"
 COM_JOOMLAUPDATE_PREUPDATE_CHECK_CAPTION="Table of server settings to check before update."
 COM_JOOMLAUPDATE_PREUPDATE_HEADING_CHECKED="Checked"
 COM_JOOMLAUPDATE_PREUPDATE_HEADING_REQUIREMENT="Requirement"
+COM_JOOMLAUPDATE_PREUPDATE_UNKNOWN_EXTENSION_MANIFESTCACHE_VERSION="Unknown version"
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.com_joomlaupdate.ini]: Where does this string get displayed - might need to be Capitalised

Comment (zero-24) [reply administrator/language/en-GB/en-GB.com_joomlaupdate.ini]: image

In the Pre-Update checker under the extension part. commit has been made, Thanks! :)

Comment (brianteeman) [reply administrator/language/en-GB/en-GB.com_joomlaupdate.ini]: thanks

File: administrator/components/com_joomlaupdate/Model/UpdateModel.php [hunk 2]
```diff
@@ -1354,7 +1354,7 @@ public function getNonCoreExtensions()
 		{
 			$decode = json_decode($extension->manifest_cache);
 			$this->translateExtensionName($extension);
-			$extension->version = $decode->version;
+			$extension->version = isset($decode->version) ? $decode->version : Text::_('COM_JOOMLAUPDATE_UNKNOWN_EXTENSION_MANIFESTCACHE_VERSION');
```
Comment (infograf768) [inline administrator/components/com_joomlaupdate/Model/UpdateModel.php]: Wrong constant:
Should be:
`COM_JOOMLAUPDATE_PREUPDATE_UNKNOWN_EXTENSION_MANIFESTCACHE_VERSION`

Comment (roland-d) [reply administrator/components/com_joomlaupdate/Model/UpdateModel.php]: Correct because now you get:
image

Comment (zero-24) [reply administrator/components/com_joomlaupdate/Model/UpdateModel.php]: Thanks fixed by the last commit. I have changed the constant just before committing but only at one place :D

**Comment by richard67** [comment]
I have tested this item :white_check_mark: successfully on 36dd0e7ad1bb81e824b1f7485643babc0cf4d906<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26293">issues.joomla.org/tracker/joomla-cms/26293</a>.</sub>

**Comment by Quy** [comment]
Will you do another PR for the following?

`PHP Notice:  Trying to get property 'version' of non-object in \administrator\components\com_installer\Model\DatabaseModel.php on line 490`

File: administrator/components/com_joomlaupdate/Model/UpdateModel.php [hunk 3]
```diff
@@ -1354,7 +1354,7 @@ public function getNonCoreExtensions()
 		{
 			$decode = json_decode($extension->manifest_cache);
 			$this->translateExtensionName($extension);
-			$extension->version = $decode->version;
+			$extension->version = isset($decode->version) ? $decode->version : Text::_('COM_JOOMLAUPDATE_PREUPDATE_UNKNOWN_EXTENSION_MANIFESTCACHE_VERSION');
```
Comment (Quy) [inline administrator/components/com_joomlaupdate/Model/UpdateModel.php]:
```suggestion
			$extension->version = $decode->version ?? Text::_('COM_JOOMLAUPDATE_PREUPDATE_UNKNOWN_EXTENSION_MANIFESTCACHE_VERSION');
```

This should work too?

**Comment by infograf768** [comment]
@Quy 
How to reproduce that error for DatabaseModel ?

**Comment by infograf768** [comment]
@zero-24 
We have also
```
Notice: Undefined property: stdClass::$version in /Applications/MAMP/htdocs/installmulti/joomla40/administrator/components/com_installer/tmpl/manage/default.php on line 108

Notice: Undefined property: stdClass::$version in /Applications/MAMP/htdocs/installmulti/joomla40/administrator/components/com_installer/tmpl/manage/default.php on line 124

```
This shows in Extensions=>Manage for the Sample Data Testing plugin

**Comment by infograf768** [comment]
These means that we may need a more general string for "Unknown version" as it can be used with exactly the same value in com_installer, including for this PR. Maybe `JGLOBAL_UNKNOWN_VERSION="Unknown Version"`

**Comment by Quy** [comment]
@infograf768 Go to System > Update > Joomla > Pre-Update Check

**Comment by zero-24** [comment]
Well the deeper problem is why is the manifest cache not up-to-date, has this been reproduced in a live site too? This is just a hot fix not fixing the real cause of the problem.

**Comment by Eriiond** [comment]
I have tested this item :white_check_mark: successfully on 36dd0e7ad1bb81e824b1f7485643babc0cf4d906<br /><hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26293">issues.joomla.org/tracker/joomla-cms/26293</a>.</sub>

**Comment by vijaykhollam** [comment]
I have tested this item :white_check_mark: successfully on e44cf7a6b680d13cdec51141a0a7ec8fda220cff
#pbf19 #test1
<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26293">issues.joomla.org/tracker/joomla-cms/26293</a>.</sub>

**Comment by wilsonge** [comment]
Can you fix conflicts please?

**Comment by richard67** [comment]
@zero-24 It just needs to rename file `administrator/language/en-GB/en-GB.com_joomlaupdate.ini` to  `administrator/language/en-GB/com_joomlaupdate.ini`.

**Comment by zero-24** [comment]
Done @wilsonge thanks @richard67

**Comment by wilsonge** [comment]
Thanks!

**Pull request merged**